drivers/hmis/lib/hmis_data_cleanup/util.rb

```diff
@@ -155,6 +173,41 @@ def self.delete_duplicate_exit_records!
       GrdaWarehouse::Hud::Exit.where(id: ids_to_delete).update_all(DateDeleted: Time.current, source_hash: nil)
     end
 
+    # This cleanup routine was added because duplicate EnrollmentIDs were leading to uniqueness violation constraints
+    # when Client records were merged.
+    #
+    # 1) Hard-delete deleted enrollments where there is any NON-deleted Enrollment that shares an EnrollmentID
+    # 2) Hard-delete deleted all but 1 enrollment where there is a set of deleted Enrollments that share an EnrollmentID
+    def self.hard_delete_duplicate_deleted_enrollments!
+      data_source_id = GrdaWarehouse::DataSource.hmis.first.id
+      result = GrdaWarehouse::Hud::Enrollment.with_deleted.
```

Comment:
This seems like a really complex way to do this.  Could we not just do something like?
```
open_enrollments = GrdaWarehouse::Hud::Enrollment.where(data_source_id: data_source_id).pluck(:enrollment_id)
deleted_enrollments = GrdaWarehouse::Hud::Enrollment.only_deleted.where(data_source_id: data_source_id).pluck(:enrollment_id)

duplicates = open_enrollments & deleted_enrollments

GrdaWarehouse::Hud::Enrollment.only_deleted.where(data_source_id: data_source_id, enrollment_id: duplicates).delete_all
```

That's untested, but I _think_ it's roughly equivalent.

Reply:
I think thats equivalent, except for the fact that it will dedup even in cases where the ProjectID doesn't match. (but that could easily be included into your approach).  I haven't seen cases like that (same enrollment ID but diff project id) but it seems like something we'd want to retain.

It also isn't equivalent in that it retains existing duplicates among deleted enrollments. but again, that could be incorporated with this array approach.

I like your approach for readability but can we pluck all the Enrollment IDs in the whole data source?   That seemed like a lot. It's only a one-time thing, so maybe it's fine..

Reply:
Yeah, I didn't account for ProjectID, we probably should.  And good point about not checking for duplicates within the deleted records.  

I don't think plucking all the ids will be an issue.

Either approach is fine with me, I was just having some trouble following yours.

Reply:
I was having trouble writing and following it too tbh... I will update to something more readable.

Reply:
I'm having a hard time making this more readable using pluck keys, when project IDs and deduping deleted ones come into play. It feels a lot harder to reason about because we don't have access to the database Ids.



```
      active_enrollments = GrdaWarehouse::Hud::Enrollment.where(data_source_id: data_source_id).pluck(:enrollment_id, :project_id)
      deleted_enrollments = GrdaWarehouse::Hud::Enrollment.only_deleted.where(data_source_id: data_source_id).pluck(:enrollment_id, :project_id)

      duplicates = active_enrollments & deleted_enrollments
      duplicates = duplicates.to_set

      # Hard-delete deleted enrollments where there is any NON-deleted Enrollment that shares the same EnrollmentID+ProjectID
      GrdaWarehouse::Hud::Enrollment.only_deleted.
        where(data_source_id: data_source_id, enrollment_id: duplicates.map(&:first)).
        select { |e| duplicates.include?([e.enrollment_id, e.project_id]) }.
        each(&:really_destroy!)

      # deleted enrollments that are duplicated, and DON'T match existing enrollments
      deleted_enrollments_dups = deleted_enrollments.group_by(&:itself).
        filter_map { |_, v| v if v.size > 1 } - active_enrollments

      # now need to delete only the dups of the above (leave 1)
      # todo
```